src/lib/homestore_backend/gc_manager.cpp

```diff
@@ -0,0 +1,360 @@
+
+    auto defrag_blk_num = chunk->get_defrag_nblks();
+    auto total_blk_num = chunk->get_total_blks();
+
+    // defrag_blk_num > (GC_THRESHOLD_PERCENT/100) * total_blk_num, to avoid floating point number calculation
+    // TODO: avoid overflow here.
+    return 100 * defrag_blk_num > total_blk_num * GC_GARBAGE_RATE_THRESHOLD;
+}
+
+void GCManager::scan_chunks_for_gc() {
+    for (const auto& [_, chunk] : m_chunk_selector->get_all_chunks()) {
+        if (is_eligible_for_gc(chunk)) {
+            auto pdev_id = chunk->get_pdev_id();
+            auto it = m_pdev_gc_actors.find(pdev_id);
+            if (it != m_pdev_gc_actors.end()) {
+                auto& actor = it->second;
+                auto [promise, _ /* future */] = folly::makePromiseContract< bool >();
+                // TODO:: use the future to check if the normal task is completed when necessary
+                actor->add_gc_task(
+                    task_priority::normal,
+                    {chunk->get_chunk_id(), static_cast< uint8_t >(task_priority::normal), std ::move(promise)});
+            }
+        }
+    }
+}
+
+/* pdev_gc_actor */
+
+GCManager::pdev_gc_actor::pdev_gc_actor(uint32_t pdev_id, std::shared_ptr< HeapChunkSelector > chunk_selector,
+                                        std::shared_ptr< GCBlobIndexTable > index_table, HSHomeObject* homeobject) :
+        m_pdev_id{pdev_id},
+        m_chunk_selector{chunk_selector},
+        // we have a queue for each priority, so the size is the same as the number of priorities
+        m_gc_task_queue{static_cast< size_t >(task_priority::priority_count)},
+        m_reserved_chunk_queue{RESERVED_CHUNK_NUM_PER_PDEV},
+        m_index_table{index_table},
+        m_hs_home_object{homeobject} {
+
+    RELEASE_ASSERT(index_table, "index_table for a gc_actor should not be nullptr!!!");
+}
+
+void GCManager::pdev_gc_actor::start() {
+    bool stopped = true;
+    if (!m_is_stopped.compare_exchange_strong(stopped, false, std::memory_order_release, std::memory_order_relaxed)) {
+        LOGERROR("pdev gc actor for pdev_id={} is already started, no need to start again!", m_pdev_id);
+        return;
+    }
+    // thread number is the same as reserved chunk, which can make sure every gc thread can take a reserved chunk
+    // for gc
+    m_gc_executor = std::make_shared< folly::IOThreadPoolExecutor >(RESERVED_CHUNK_NUM_PER_PDEV);
+    for (uint32_t i = 0; i < RESERVED_CHUNK_NUM_PER_PDEV; i++) {
+        m_gc_executor->add([this] {
+            while (!m_is_stopped.load(std::memory_order_relaxed)) {
+                GCManager::gc_task task;
+                if (m_gc_task_queue.try_dequeue(task)) {
+                    process_gc_task(task);
+                } else {
+                    // if we use a condition variable(cv), when stop, the cv might lost the notification. so, it`s
+                    // better to do sleep and check
+                    LOGINFO("no gc task found, sleep 1s!");
```

Comment:
@xiaoxichen I recall why I use this busy loop instead of a condition variable notification.

when we stop, we set the stop flag and notify the gc thread. but if the gc thread is working on gc and not cv.wait(), it will lose this notification and never be notified again. we need a more complex mechanism to achieve. I can later use a seperate PR to do this .